File: scala-advent-of-code/src/main/scala/advent/twenty_sixteen/Day02BathroomSecurity.scala
package advent.twenty_sixteen

import advent.common.DailyProblem
import advent.utilities.{FileUtilities, Point, TurnLeft, TurnRight}

abstract class KeypadMove
case class KeypadMoveUp() extends KeypadMove
case class KeypadMoveDown() extends KeypadMove
case class KeypadMoveLeft() extends KeypadMove
case class KeypadMoveRight() extends KeypadMove

class Day02BathroomSecurity(filename: String) extends DailyProblem[String, String] {
  private val keypadPart1: Array[Array[Char]] = Array[Array[Char]]( //
    Array('*', '*', '*', '*', '*'), //
    Array('*', '1', '2', '3', '*'), //
    Array('*', '4', '5', '6', '*'), //
    Array('*', '7', '8', '9', '*'), //
    Array('*', '*', '*', '*', '*'))

  private val keypadPart2: Array[Array[Char]] = Array[Array[Char]](
    Array('*', '*', '*', '*', '*', '*', '*'), //
    Array('*', '*', '*', '1', '*', '*', '*'), //
    Array('*', '*', '2', '3', '4', '*', '*'), //
    Array('*', '5', '6', '7', '8', '9', '*'), //
    Array('*', '*', 'A', 'B', 'C', '*', '*'), //
    Array('*', '*', '*', 'D', '*', '*', '*'), //
    Array('*', '*', '*', '*', '*', '*', '*'))

  private def parser(line: String): Seq[KeypadMove] = line.map {
    case 'U' => KeypadMoveUp()
    case 'D' => KeypadMoveDown()
    case 'L' => KeypadMoveLeft()
    case 'R' => KeypadMoveRight()
    case c => throw new IllegalArgumentException("Unknown character " + c)
  }

  private val moves = FileUtilities.readFile(filename, parser)

  def check(keypad: Array[Array[Char]], oldPoint: Point, newPoint: Point): Point = {
    if (keypad(newPoint.y)(newPoint.x) == '*') oldPoint else newPoint
  }

  def walkMoves(moves: Seq[Seq[KeypadMove]], keypad: Array[Array[Char]], fivePos: Point): String = {

    def line(remaining: Seq[KeypadMove], current: Point): (Char, Point) = {
      val p = remaining.foldLeft(current)((acc, move) => {
        move match {
          case KeypadMoveUp() => check(keypad, acc, Point(acc.y - 1, acc.x))
          case KeypadMoveDown() => check(keypad, acc, Point(acc.y + 1, acc.x))
          case KeypadMoveLeft() => check(keypad, acc, Point(acc.y, acc.x - 1))
          case KeypadMoveRight() => check(keypad, acc, Point(acc.y, acc.x + 1))
        }
      })
      (keypad(p.y)(p.x), p)
    }

    moves.foldLeft[(String, Point)](("", fivePos))((acc, move) => {
      val res = line(move, acc._2)
      (acc._1 + res._1, res._2)
    })._1
  }

  override val part1Answer: String = walkMoves(moves, keypadPart1, Point(2, 2))
  override val part2Answer: String = walkMoves(moves, keypadPart2, Point(3, 1))
}

File: java-advent-of-code/src/main/java/advent/twenty_sixteen/Day02BathroomSecurity.java
package advent.twenty_sixteen;

import advent.common.DailyProblem;
import advent.utilities.FileUtilities;
import advent.utilities.Parsers;

import java.util.List;

public class Day02BathroomSecurity implements DailyProblem<String, String> {
    private final String part1Answer;
    private final String part2Answer;

    public Day02BathroomSecurity(String filename) {
        List<String> lines = FileUtilities.readLines(filename, Parsers::TO_STRING);

        char[][] keypadPart1 = new char[][]{ //
                {'*', '*', '*', '*', '*'}, //
                {'*', '1', '2', '3', '*'}, //
                {'*', '4', '5', '6', '*'}, //
                {'*', '7', '8', '9', '*'}, //
                {'*', '*', '*', '*', '*'}  //
        };


        char[][] keypadPart2 = new char[][]{
                {'*', '*', '*', '*', '*', '*', '*'}, //
                {'*', '*', '*', '1', '*', '*', '*'}, //
                {'*', '*', '2', '3', '4', '*', '*'}, //
                {'*', '5', '6', '7', '8', '9', '*'}, //
                {'*', '*', 'A', 'B', 'C', '*', '*'}, //
                {'*', '*', '*', 'D', '*', '*', '*'}, //
                {'*', '*', '*', '*', '*', '*', '*'}  //
        };

        this.part1Answer = walkKeypad(lines, 2, 2, keypadPart1);
        this.part2Answer = walkKeypad(lines, 3, 1, keypadPart2);
    }

    private String walkKeypad(List<String> lines, int r, int c, char[][] keypad) {
        String code = "";
        for (String line : lines) {
            for (char chr : line.toCharArray()) {
                int pr = r;
                int pc = c;
                switch (chr) {
                    case 'U':
                        r--;
                        break;
                    case 'D':
                        r++;
                        break;
                    case 'L':
                        c--;
                        break;
                    case 'R':
                        c++;
                        break;
                    default:
                        throw new IllegalArgumentException("Unknown step " + chr);
                }
                if (keypad[r][c] == '*') {
                    r = pr;
                    c = pc;
                }
            }
            code += keypad[r][c];
        }
        return code;
    }

    @Override
    public String getPart1Answer() {
        return part1Answer;
    }

    @Override
    public String getPart2Answer() {
        return part2Answer;
    }
}
